src/olympia/amo/models.py

```diff
@@ -505,3 +507,55 @@ class LongNameIndex(models.Index):
     """Django's Index, but with a longer allowed name since we don't care about
     compatibility with Oracle."""
     max_name_length = 64  # Django default is 30, but MySQL can go up to 64.
+
+
+class FilterableManyToManyDescriptor(ManyToManyDescriptor):
+
+    def __init__(self, *args, **kwargs):
+        self.q_filter = kwargs.pop('q_filter', None)
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def _get_manager_with_default_filtering(cls, manager, q_filter):
+        class ManagerWithFiltering(manager):
+            def get_queryset(self):
+                try:
+                    return self.instance._prefetched_objects_cache[
+                        self.prefetch_cache_name]
+                except (AttributeError, KeyError):
+                    pass
+                qs = super().get_queryset()
+                return qs.filter(q_filter or models.Q())
+
+        return ManagerWithFiltering
+
+    @cached_property
+    def related_manager_cls(self):
+        cls = super().related_manager_cls
+        return self._get_manager_with_default_filtering(cls, self.q_filter)
+
+
+class FilterableManyToManyField(models.fields.related.ManyToManyField):
+
```

Comment:
Given that this code is inheriting from undocumented classes like `ManyToManyDescriptor` and overriding many undocumented internal methods we need a lot more documentation and tests. Docstrings to explain what the classes are and why they exists, comments in overridden methods helping understand what's different compared to django's, and some tests that are isolated from the rest that explicitly test even basic behavior from django is still working. Otherwise this will be a nightmare during the next django upgrade in a similar way translations app always is.

Reply:
https://github.com/mozilla/addons-server/pull/14827/commits/ba78d4f9efcc0b591ebe6fe123e65c12103b8be9